vault/logical_system.go

```diff
@@ -1406,6 +1456,210 @@ func (b *SystemBackend) handleRotate(
 	return nil, nil
 }
 
+func (b *SystemBackend) handleWrappingWrap(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	if req.WrapTTL == 0 {
+		return logical.ErrorResponse("endpoint requires response wrapping to be used"), logical.ErrInvalidRequest
+	}
+
+	return &logical.Response{
+		Data: data.Raw,
+	}, nil
+}
+
+func (b *SystemBackend) handleWrappingUnwrap(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	// If a third party is unwrapping (rather than the calling token being the
+	// wrapping token) we detect this so that we can revoke the original
+	// wrapping token after reading it
+	var thirdParty bool
+
+	token := data.Get("token").(string)
+	if token != "" {
+		thirdParty = true
+	} else {
+		token = req.ClientToken
+	}
+
+	if thirdParty {
+		defer b.Core.tokenStore.Revoke(token)
+	}
+
+	cubbyReq := &logical.Request{
+		Operation:   logical.ReadOperation,
+		Path:        "cubbyhole/response",
+		ClientToken: token,
+	}
+	cubbyResp, err := b.Core.router.Route(cubbyReq)
+	if err != nil {
+		return logical.ErrorResponse(fmt.Sprintf("error looking up wrapping information: %v", err)), nil
+	}
+	if cubbyResp == nil {
+		return logical.ErrorResponse("no information found; wrapping token may be from a previous Vault version"), nil
```

Comment:
Yeah, I'm unsure about this. The reason I think it's better as an error is that if the response is success they may be expecting to be able to unwrap (especially if it's not a human, but code looking at error returns). So this ensures that they don't think they have valid data and attempt to proceed.